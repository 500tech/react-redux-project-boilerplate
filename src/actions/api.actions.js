// @flow
import { set } from 'lodash/fp';
import type { ApiAction } from 'types/redux.types';

export const apiAction = (action: ApiAction): ApiAction =>
  set('meta.api', true, action);
