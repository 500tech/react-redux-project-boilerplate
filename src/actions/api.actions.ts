import { BaseAction } from 'types/base-redux.types';

export type HttpMethod =
  | 'get'
  | 'post'
  | 'put'
  | 'delete'
  | 'path'
  | 'head'
  | 'options';

type OnSuccess<ResponseBody> = (data: ResponseBody) => BaseAction;
type OnError = (error: any) => void;

export interface ApiAction<ResponseBody> extends BaseAction {
  meta: {
    api: true;
  };
  payload: {
    networkLabel: string;
    method: HttpMethod;
    path: string;
    data?: any;
    baseUrl?: string;
    onSuccess: OnSuccess<ResponseBody> | OnSuccess<ResponseBody>[];
    onError?: OnError | OnError[];
  };
}
