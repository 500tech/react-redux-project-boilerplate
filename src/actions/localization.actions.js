// @flow
// TODO: remove this file if no localization is needed
import type { BaseAction } from 'types/redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export type SetLocaleAction = BaseAction<{|
  +locale: string
|}>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
