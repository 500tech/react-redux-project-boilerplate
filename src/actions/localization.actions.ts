// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleActionPayload {
  locale: string;
}

export type SetLocaleAction = BaseAction<SetLocaleActionPayload>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
