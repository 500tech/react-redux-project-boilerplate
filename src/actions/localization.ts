// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux';
import { LocaleTypes } from 'constants/locales';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleAction extends BaseAction {
  payload: {
    locale: string;
  };
}

export const setLocale = (locale: LocaleTypes) => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
