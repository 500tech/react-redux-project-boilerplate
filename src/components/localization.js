// @flow
import * as React from 'react';
import forEach from 'lodash/forEach';
import { connect } from 'react-redux';
import { IntlProvider, addLocaleData } from 'react-intl';
import moment from 'moment';

import store from 'store';

import locales from 'constants/locales';

import type { State } from 'types/redux.types';
import type { MapStateToProps } from 'react-redux';

forEach(locales, (locale, key) =>
  moment.defineLocale(key, locale.dateTimeFormat)
);

// Go over all of the available locales and register them
forEach(locales, (value, key) => {
  addLocaleData({
    locale: key,
    // Couldn't find any documentation about 'pluralRuleFunction', throws error if not present
    pluralRuleFunction: () => {}
  });
});

type ConnectedProps = {
  locale: string
};

type OwnProps = {
  children: React.Node
};

export const Localization = ({
  locale,
  children
}: ConnectedProps & OwnProps) => (
  <IntlProvider
    locale={locale}
    key={locale}
    messages={locales[locale].translations}>
    {children}
  </IntlProvider>
);

const mapStateToProps: MapStateToProps<State, OwnProps, {}> = ({
  localization
}: State) => ({
  locale: localization.locale
});

let currentLocale = store.getState().localization.locale;

moment.locale(currentLocale);
store.subscribe(() => {
  const newLocale = store.getState().localization.locale;

  if (newLocale !== currentLocale) {
    currentLocale = newLocale;
    moment.locale(currentLocale);
  }
});

export default connect(mapStateToProps, {})(Localization);
