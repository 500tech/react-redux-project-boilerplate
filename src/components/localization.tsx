import * as React from 'react';
import forEach from 'lodash/forEach';
import { connect } from 'react-redux';
import { IntlProvider, addLocaleData, Locale } from 'react-intl';
import moment, { LocaleSpecification } from 'moment';

import store from 'store';

import locales, { LocaleTypes } from 'constants/locales';

import { State } from 'types/redux.types';

forEach(locales, (locale, key) =>
  moment.defineLocale(
    key,
    (locale.dateTimeFormat as unknown) as LocaleSpecification
  )
);

// Go over all of the available locales and register them
forEach(locales, (_value, key: string) => {
  addLocaleData(({
    locale: key,
    // Couldn't find any documentation about 'pluralRuleFunction', throws error if not present
    pluralRuleFunction: () => {}
  } as unknown) as Locale);
});

export const Localization = ({ locale, children }: Props) => (
  <IntlProvider
    locale={locale}
    key={locale}
    messages={locales[locale].translations}>
    {children}
  </IntlProvider>
);

interface OwnProps {
  children: React.ReactNode;
}

interface StateProps {
  locale: LocaleTypes;
}

type Props = StateProps & OwnProps;

const mapStateToProps = (state: State): StateProps => ({
  locale: state.localization.locale
});

let currentLocale = (store.getState() as State).localization.locale;

moment.locale(currentLocale);
store.subscribe(() => {
  const newLocale = (store.getState() as State).localization.locale;

  if (newLocale !== currentLocale) {
    currentLocale = newLocale;
    moment.locale(currentLocale);
  }
});

export default connect(mapStateToProps)(Localization);
