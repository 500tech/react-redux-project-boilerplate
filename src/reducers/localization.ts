// TODO: remove this file if no localization is needed
import { get, set } from 'lodash/fp';
import { handleActions } from 'redux-actions';

import { setLocale } from 'actions/localization';

import * as AT from 'actions/localization';
import { LocaleTypes } from 'constants/locales';

export interface LocalizationState {
  locale: LocaleTypes;
}

const initialState: LocalizationState = {
  locale: 'en-US'
};

const localization = handleActions<LocalizationState>(
  {
    [AT.SET_LOCALE]: (
      state: LocalizationState,
      action: ReturnType<typeof setLocale>
    ): LocalizationState => {
      const locale = get('payload.locale', action);

      return set('locale', locale, state);
    }
  },
  initialState
);

export default localization;
