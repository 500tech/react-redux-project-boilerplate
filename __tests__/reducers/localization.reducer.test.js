import localizationReducer from 'reducers/localization.reducer';
import { setLocale } from 'actions/localization.actions';

describe('Reducer: Localization', () => {
  test('should set locale', () => {
    const action = setLocale('en-US');
    const result = localizationReducer(undefined, action);

    expect(result).toMatchSnapshot();
  });
});
