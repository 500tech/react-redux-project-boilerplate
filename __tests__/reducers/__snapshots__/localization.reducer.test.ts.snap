// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer: Localization should set locale 1`] = `
Object {
  "locale": "en-US",
}
`;
